lib/webdriveragent.js

```diff
@@ -46,123 +48,210 @@ class WebDriverAgent {
+          // passed before we get to the real startup logs, otherwise
+          // we expect two lines, one after another
+          //     Jul 20 13:03:57 iamPhone XCTRunner[296] <Warning>: Built at Jul 20 2016 13:03:50
+          //     Jul 20 13:03:57 iamPhone XCTRunner[296] <Warning>: ServerURLHere->http://10.35.4.122:8100<-ServerURLHere
+          if (!reachedEnd) {
+            let dateMatch = LOG_STARTTIME_REGEX.exec(stdout);
+            if (dateMatch) {
+              let buildTime = Date.parse(dateMatch[1]);
+              if (buildTime > startTime) {
+                reachedEnd = true;
+              }
+            }
+          }
 
-      this.url = url.parse(agentUrl);
+          if (reachedEnd) {
+            let match = AGENT_STARTED_REGEX.exec(stdout);
+            if (match) {
+              agentUrl = match[1];
+              log.info(`Detected that WebDriverAgent is running at url '${agentUrl}'`);
+              if (!agentUrl) {
+                log.errorAndThrow(new Error('No url detected from WebDriverAgent'));
+              }
+              // check for (null) url (e.g., 'http://(null):8100'), which happens on real devices
+              if (/(\(null\))/.test(agentUrl)) {
+                let msg = `Unable to detect url from WebDriverAgent: ${agentUrl}. ` +
+                          'Is the device on the same network as the server?';
+                log.errorAndThrow(new Error(msg));
+              }
 
-      this.jwproxy = new JWProxy({host: this.url.hostname, port: this.url.port, base: ''});
-      this.jwproxy.sessionId = sessionId;
-      this.proxyReqRes = this.jwproxy.proxyReqRes.bind(this.jwproxy);
+              return true;
+            }
+          }
 
-      return resolve(url);
+          // periodically log, so it does not look like everything died
+          if (lineCount % 100 === 0) {
+            log.debug('Waiting for WebDriverAgent server to finish loading...');
+          }
+
+          return false;
+        });
+
+        log.info('Waiting for WebDriverAgent to start on device');
+        await agentStartedOnDevice;
+        log.info(`WebDriverAgent started at url '${agentUrl}'`);
+
+        this.url = url.parse(agentUrl);
+        this.jwproxy = new JWProxy({server: this.url.hostname, port: this.url.port, base: ''});
+        this.jwproxy.sessionId = sessionId;
+        this.proxyReqRes = this.jwproxy.proxyReqRes.bind(this.jwproxy);
+        return resolve(url);
+      } catch (err) {
+        log.warn(`Starting xcodebuild failed: ${err.message}`);
+        reject(err);
+      }
     });
   }
 
   async checkForDependencies () {
-    if (!await fs.hasAccess(this.bootstrapPath + "/Carthage")) {
+    if (!await fs.hasAccess(`${this.bootstrapPath}/Carthage`)) {
       await exec('/bin/bash', ['Scripts/bootstrap.sh'], {cwd: this.bootstrapPath});
     }
   }
 
-  createXcodeBuildSubProcess () {
+  createXcodeBuildSubProcess (reject, realDevice = false) {
```

Comment:
I have to say I'm not a big fan of this pattern of passing `reject` around, and feel like there must be a more transparent way...